File: ga/glm4_result_2024_6/result/weakptr_l_glm/3_weakptr_l.cpp
use std::rc::{Rc, Weak};
use std::cell::RefCell;

struct MyStruct {
    value: i32,
}

impl MyStruct {
    fn new(value: i32) -> Rc<RefCell<MyStruct>> {
        Rc::new(RefCell::new(MyStruct { value }))
    }

    fn observe(weak: &Weak<RefCell<MyStruct>>) {
        if let Some(shared) = weak.upgrade() {
            println!("Object is alive, value: {}", shared.borrow().value);
        } else {
            println!("Object has been destroyed");
        }
    }
}

fn main() {
    let my_struct = MyStruct::new(10);
    let weak_my_struct = Rc::downgrade(&my_struct);

    {
        let my_struct = Rc::clone(&my_struct);
        println!("Inside inner scope");
        MyStruct::observe(&weak_my_struct);

        my_struct.borrow_mut().value = 20;
        println!("Value: {}", my_struct.borrow().value);
    }

    println!("Outside inner scope");
    MyStruct::observe(&weak_my_struct);

    drop(my_struct);
    MyStruct::observe(&weak_my_struct);
}

File: ga/4omini_result_2024_7/result/weakptr_l_mini/3_weakptr_l.rs
use std::rc::{Rc, Weak};

struct MyStruct {
    value: i32,
}

impl MyStruct {
    fn new(v: i32) -> MyStruct {
        println!("Constructor");
        MyStruct { value: v }
    }
}

impl Drop for MyStruct {
    fn drop(&mut self) {
        println!("Destructor");
    }
}

fn observe(weak_ptr: &Weak<MyStruct>) {
    if let Some(shared_ptr) = weak_ptr.upgrade() {
        println!("Object is alive, value: {}", shared_ptr.value);
    } else {
        println!("Object has been destroyed");
    }
}

fn main() {
    let ptr1 = Rc::new(MyStruct::new(10));
    let weak_ptr1 = Rc::downgrade(&ptr1);

    {
        let ptr2 = Rc::clone(&ptr1);
        println!("Inside inner scope");
        observe(&weak_ptr1);

        let weak_ptr2 = Rc::downgrade(&ptr2);
        if let Some(shared_ptr) = weak_ptr2.upgrade() {
            shared_ptr.value = 20;
            println!("Value: {}", shared_ptr.value);
        }
    } // ptr2 goes out of scope here

    println!("Outside inner scope");
    observe(&weak_ptr1);

    // Resetting ptr1 to drop the reference and trigger destructor
    drop(ptr1);
    observe(&weak_ptr1);
}